### ISSUE TITLE
fix: add c:autoTitleDeleted element to new charts

### ISSUE THREAD

**Issue by karakanb** [scanny/python-pptx#460]
Hi,

I have been building some PPTX files using the library, and everything works good when I am using Microsoft Office; however, today I have switched on to LibreOffice, and it turns out the chart titles for bar chart and pie chart are not rendered, while line chart titles are rendered. The full reproducible piece of code resides in a big project, though all of the individual components are created using the same classes. **The weird thing is, if I open the file with PowerPoint when I first download it and save it, then LibreOffice can read the charts appropriately, which probably means the resulting file that I build has some wrong structure and PowerPoint fixes it.** The code below is used for creating and returning the individual components, which sets the required attributes for the charts:

### Line Chart
```python
chart_data = ChartData()
chart_data.categories = self.data['dimension']

# Define chart data ---------------------
for series, label in self.labels['series'].items():
    chart_data.add_series(label, self.data[series])

# Create chart
graphic_frame = slide.placeholders[idx].insert_chart(XL_CHART_TYPE.LINE, chart_data)
chart = graphic_frame.chart

# Chart text formatting
chart.font.size = Pt(12)
chart.font.name = 'Calibri Light '

# Chart title
chart.chart_title.text_frame.text = self.title

chart.has_legend = True
chart.legend.position = XL_LEGEND_POSITION.BOTTOM
chart.legend.include_in_layout = False
```

### Pie Chart
```python
chart_data = ChartData()
chart_data.categories = self.data['colour']

# define chart data
chart_data.add_series(self.labels['metric'], self.data['x'])

# Create chart
graphic_frame = slide.placeholders[idx].insert_chart(XL_CHART_TYPE.PIE, chart_data)
chart = graphic_frame.chart

# Chart text formatting
chart.font.size = Pt(12)
chart.font.name = 'Calibri Light '

# Chart title
chart.has_title = True
chart.chart_title.has_text_frame = True
chart.chart_title.text_frame.text = self.title

# Data labels
plot = chart.plots[0]
plot.has_data_labels = True
data_labels = plot.data_labels
data_labels.position = XL_LABEL_POSITION.BEST_FIT

chart.has_legend = True
chart.legend.position = XL_LEGEND_POSITION.RIGHT
chart.legend.include_in_layout = False
```

### Bar Chart
```python
chart_data = ChartData()
chart_data.categories = self.data['dimension']

# Define chart data ---------------------
for series, label in self.labels['series'].items():
    chart_data.add_series(label, self.data[series])

# Create chart
graphic_frame = slide.placeholders[idx].insert_chart(XL_CHART_TYPE.COLUMN_CLUSTERED, chart_data)
chart = graphic_frame.chart

# Chart text formatting
chart.font.size = Pt(12)
chart.font.name = 'Calibri Light '

# Chart title
chart.chart_title.text_frame.text = self.title

chart.has_legend = True
chart.legend.position = XL_LEGEND_POSITION.BOTTOM
chart.legend.include_in_layout = False
```

## Image Sample
The following is the screenshot from the same pptx opened in Microsoft PowerPoint:
<img width="468" alt="screen shot 2018-11-12 at 22 55 28" src="https://user-images.githubusercontent.com/16530606/48380812-89dc2d80-e6d9-11e8-949c-09ab9618f812.png">

This one is the screenshot of the same pie chart from LibreOffice:
<img width="468" alt="screen shot 2018-11-12 at 23 03 48" src="https://user-images.githubusercontent.com/16530606/48380833-9c566700-e6d9-11e8-9216-11cfc49f8f65.png">

To be honest this package has enabled us a lot to create very useful stuff, thank you very much for that @scanny, right now this issue is blocking us because we are trying to render PDF out of the PPTX and unoconv seems to be the only option there; however, with this problem the resulting PDF becomes useless. Any kind of help is appreciated here, I can also share the master template. 

To debug the issue, I have extracted the file I have built before I open the file with PowerPoint into a seperate directory and initialized a Git repo in there and committed all the files in there. Then I opened the file with PowerPoint, saved as a different file, and extracted this new file to another folder. In this case, when I moved the `.git` folder from the first directory to the second directory and run `git status` there, the result shows almost all of the files as changed:

```bash
karakanb@karakanb:~/resaved (master *)$ git status
On branch master
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   [Content_Types].xml
	modified:   _rels/.rels
	modified:   docProps/app.xml
	modified:   docProps/core.xml
	modified:   ppt/_rels/presentation.xml.rels
	modified:   ppt/charts/_rels/chart1.xml.rels
	modified:   ppt/charts/_rels/chart2.xml.rels
	modified:   ppt/charts/_rels/chart3.xml.rels
	modified:   ppt/charts/_rels/chart4.xml.rels
	modified:   ppt/charts/_rels/chart5.xml.rels
	modified:   ppt/charts/_rels/chart6.xml.rels
	modified:   ppt/charts/chart1.xml
	modified:   ppt/charts/chart2.xml
	modified:   ppt/charts/chart3.xml
	modified:   ppt/charts/chart4.xml
	modified:   ppt/charts/chart5.xml
	modified:   ppt/charts/chart6.xml
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet1.xlsx
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet2.xlsx
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet3.xlsx
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet4.xlsx
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet5.xlsx
	deleted:    ppt/embeddings/Microsoft_Excel_Sheet6.xlsx
	modified:   ppt/presentation.xml
	modified:   ppt/slideLayouts/_rels/slideLayout1.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout2.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout3.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout4.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout5.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout6.xml.rels
	modified:   ppt/slideLayouts/_rels/slideLayout7.xml.rels
	modified:   ppt/slideLayouts/slideLayout1.xml
	modified:   ppt/slideLayouts/slideLayout2.xml
	modified:   ppt/slideLayouts/slideLayout3.xml
	modified:   ppt/slideLayouts/slideLayout4.xml
	modified:   ppt/slideLayouts/slideLayout5.xml
	modified:   ppt/slideLayouts/slideLayout6.xml
	modified:   ppt/slideLayouts/slideLayout7.xml
	modified:   ppt/slideMasters/_rels/slideMaster1.xml.rels
	modified:   ppt/slideMasters/slideMaster1.xml
	modified:   ppt/slides/_rels/slide1.xml.rels
	modified:   ppt/slides/_rels/slide2.xml.rels
	modified:   ppt/slides/_rels/slide3.xml.rels
	modified:   ppt/slides/_rels/slide4.xml.rels
	modified:   ppt/slides/_rels/slide5.xml.rels
	modified:   ppt/slides/_rels/slide6.xml.rels
	modified:   ppt/slides/_rels/slide7.xml.rels
	modified:   ppt/slides/_rels/slide8.xml.rels
	modified:   ppt/slides/slide1.xml
	modified:   ppt/slides/slide2.xml
	modified:   ppt/slides/slide3.xml
	modified:   ppt/slides/slide4.xml
	modified:   ppt/slides/slide5.xml
	modified:   ppt/slides/slide6.xml
	modified:   ppt/slides/slide7.xml
	modified:   ppt/slides/slide8.xml
	modified:   ppt/viewProps.xml

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.DS_Store
	ppt/.DS_Store
	ppt/embeddings/Microsoft_Excel_Worksheet.xlsx
	ppt/embeddings/Microsoft_Excel_Worksheet1.xlsx
	ppt/embeddings/Microsoft_Excel_Worksheet2.xlsx
	ppt/embeddings/Microsoft_Excel_Worksheet3.xlsx
	ppt/embeddings/Microsoft_Excel_Worksheet4.xlsx
	ppt/embeddings/Microsoft_Excel_Worksheet5.xlsx
	ppt/embeddings/~$Microsoft_Excel_Worksheet.xlsx
```

In this case, I have checked the deleted worksheet files using the `diff` cli tool, and the file contents seem to be identical, they are just renamed, but all the other files have significant changes in them, which makes my original PPTX unusable for LibreOffice automatically. Any kind of help is appreciated, thank you.

**Comment by scanny** [comment]
I think you're on the right track with the diff, but I think we can narrow it down quite a bit. The difference will definitely be in chart1.xml if you only create a single chart. `opc-diag` allows you to diff a particular part (xml file) between two .pptx files.

So you would follow the same procedure, but limit the experiment to a single slide containing a single chart, and then compare `chart1.xml` between the fresh-from-python-pptx version and the after-saving-with-PowerPoint version, something like this:

```bash
opc diff-item before.pptx after.pptx chart1.xml
```

If you can post that diff I can help you narrow it down. I expect you'll find a difference between the two somewhere around the chart title element, which is just under the `c:chart` element. There are more details about the XML you'll find there on this analysis page: https://python-pptx.readthedocs.io/en/latest/dev/analysis/cht-chart-title.html.

**Comment by scanny** [comment]
Also, pay particular attention to the presence or absence of a `c:autoTitleDeleted` element like so:

```xml
<c:chart>
  <c:title .../>
  <c:autoTitleDeleted .../>
```

My initial hypothesis is that is the element that PowerPoint is adding in when it saves and that PowerPoint interprets a missing one of those differently than LibreOffice does.

**Comment by karakanb** [comment]
Hey, thanks for detailed answers. I have created a simple PPTX with two charts, here is the original chart XML that I've built with python-pptx:
```xml
<?xml version='1.0' encoding='UTF-8' standalone='yes'?>
<c:chartSpace xmlns:c="http://schemas.openxmlformats.org/drawingml/2006/chart" 
    xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" 
    xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships">
    <c:chart>
        <c:title>
            <c:tx>
                <c:rich>
                    <a:bodyPr/>
                    <a:lstStyle/>
                    <a:p>
                        <a:pPr>
                            <a:defRPr/>
                        </a:pPr>
                        <a:r>
                            <a:t>Users by Source</a:t>
                        </a:r>
                    </a:p>
                </c:rich>
            </c:tx>
            <c:layout/>
            <c:overlay val="0"/>
        </c:title>
        <c:plotArea>
            <c:pieChart>
                <c:varyColors val="1"/>
                <c:ser>
                    <c:idx val="0"/>
                    <c:order val="0"/>
                    <c:tx>
                        <c:strRef>
                            <c:f>Sheet1!$B$1</c:f>
                            <c:strCache>
                                <c:ptCount val="1"/>
                                <c:pt idx="0">
                                    <c:v>Users</c:v>
                                </c:pt>
                            </c:strCache>
                        </c:strRef>
                    </c:tx>
                    <c:cat>
                        <c:strRef>
                            <c:f>Sheet1!$A$2:$A$5</c:f>
                            <c:strCache>
                                <c:ptCount val="4"/>
                                <c:pt idx="0">
                                    <c:v>Instagram Stories</c:v>
                                </c:pt>
                                <c:pt idx="1">
                                    <c:v>instagram</c:v>
                                </c:pt>
                                <c:pt idx="2">
                                    <c:v>google</c:v>
                                </c:pt>
                                <c:pt idx="3">
                                    <c:v>Others</c:v>
                                </c:pt>
                            </c:strCache>
                        </c:strRef>
                    </c:cat>
                    <c:val>
                        <c:numRef>
                            <c:f>Sheet1!$B$2:$B$5</c:f>
                            <c:numCache>
                                <c:formatCode>General</c:formatCode>
                                <c:ptCount val="4"/>
                                <c:pt idx="0">
                                    <c:v>2508</c:v>
                                </c:pt>
                                <c:pt idx="1">
                                    <c:v>1063</c:v>
                                </c:pt>
                                <c:pt idx="2">
                                    <c:v>269</c:v>
                                </c:pt>
                                <c:pt idx="3">
                                    <c:v>498</c:v>
                                </c:pt>
                            </c:numCache>
                        </c:numRef>
                    </c:val>
                </c:ser>
                <c:dLbls>
                    <c:dLblPos val="bestFit"/>
                    <c:showLegendKey val="0"/>
                    <c:showVal val="1"/>
                    <c:showCatName val="0"/>
                    <c:showSerName val="0"/>
                    <c:showPercent val="0"/>
                    <c:showBubbleSize val="0"/>
                    <c:showLeaderLines val="1"/>
                </c:dLbls>
            </c:pieChart>
        </c:plotArea>
        <c:legend>
            <c:legendPos/>
            <c:overlay val="0"/>
        </c:legend>
        <c:dispBlanksAs val="gap"/>
    </c:chart>
    <c:txPr>
        <a:bodyPr/>
        <a:lstStyle/>
        <a:p>
            <a:pPr>
                <a:defRPr sz="1200">
                    <a:latin typeface="Calibri Light "/>
                </a:defRPr>
            </a:pPr>
            <a:endParaRPr lang="en-US"/>
        </a:p>
    </c:txPr>
    <c:externalData r:id="rId1">
        <c:autoUpdate val="0"/>
    </c:externalData>
</c:chartSpace>
```

Here is the one that I have opened and saved with PowerPoint:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>^M

<c:chartSpace xmlns:c="http://schemas.openxmlformats.org/drawingml/2006/chart" 
    xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" 
    xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" 
    xmlns:c16r2="http://schemas.microsoft.com/office/drawing/2015/06/chart">
    <c:date1904 val="1"/>
    <c:lang val="en-US"/>
    <c:roundedCorners val="1"/>
    <c:style val="2"/>
    <c:chart>
        <c:title>
            <c:tx>
                <c:rich>
                    <a:bodyPr/>
                    <a:lstStyle/>
                    <a:p>
                        <a:pPr>
                            <a:defRPr/>
                        </a:pPr>
                        <a:r>
                            <a:t>Users by Source</a:t>
                        </a:r>
                    </a:p>
                </c:rich>
            </c:tx>
            <c:overlay val="0"/>
        </c:title>
        <c:autoTitleDeleted val="0"/>
        <c:plotArea>
            <c:layout/>
            <c:pieChart>
                <c:varyColors val="1"/>
                <c:ser>
                    <c:idx val="0"/>
                    <c:order val="0"/>
                    <c:tx>
                        <c:strRef>
                            <c:f>Sheet1!$B$1</c:f>
                            <c:strCache>
                                <c:ptCount val="1"/>
                                <c:pt idx="0">
                                    <c:v>Users</c:v>
                                </c:pt>
                            </c:strCache>
                        </c:strRef>
                    </c:tx>
                    <c:dLbls>
                        <c:spPr>
                            <a:noFill/>
                            <a:ln>
                                <a:noFill/>
                            </a:ln>
                            <a:effectLst/>
                        </c:spPr>
                        <c:dLblPos val="bestFit"/>
                        <c:showLegendKey val="0"/>
                        <c:showVal val="1"/>
                        <c:showCatName val="0"/>
                        <c:showSerName val="0"/>
                        <c:showPercent val="0"/>
                        <c:showBubbleSize val="0"/>
                        <c:showLeaderLines val="1"/>
                        <c:extLst>
                            <c:ext uri="{CE6537A1-D6FC-4f65-9D91-7224C49458BB}" 
                                xmlns:c15="http://schemas.microsoft.com/office/drawing/2012/chart"/>
                        </c:extLst>
                    </c:dLbls>
                    <c:cat>
                        <c:strRef>
                            <c:f>Sheet1!$A$2:$A$5</c:f>
                            <c:strCache>
                                <c:ptCount val="4"/>
                                <c:pt idx="0">
                                    <c:v>Instagram Stories</c:v>
                                </c:pt>
                                <c:pt idx="1">
                                    <c:v>instagram</c:v>
                                </c:pt>
                                <c:pt idx="2">
                                    <c:v>google</c:v>
                                </c:pt>
                                <c:pt idx="3">
                                    <c:v>Others</c:v>
                                </c:pt>
                            </c:strCache>
                        </c:strRef>
                    </c:cat>
                    <c:val>
                        <c:numRef>
                            <c:f>Sheet1!$B$2:$B$5</c:f>
                            <c:numCache>
                                <c:formatCode>General</c:formatCode>
                                <c:ptCount val="4"/>
                                <c:pt idx="0">
                                    <c:v>2508</c:v>
                                </c:pt>
                                <c:pt idx="1">
                                    <c:v>1063</c:v>
                                </c:pt>
                                <c:pt idx="2">
                                    <c:v>269</c:v>
                                </c:pt>
                                <c:pt idx="3">
                                    <c:v>498</c:v>
                                </c:pt>
                            </c:numCache>
                        </c:numRef>
                    </c:val>
                    <c:extLst>
                        <c:ext uri="{C3380CC4-5D6E-409C-BE32-E72D297353CC}" 
                            xmlns:c16="http://schemas.microsoft.com/office/drawing/2014/chart">
                            <c16:uniqueId val="{00000000-0AF9-8643-ADB9-9C51A1C3652B}"/>
                        </c:ext>
                    </c:extLst>
                </c:ser>
                <c:dLbls>
                    <c:dLblPos val="bestFit"/>
                    <c:showLegendKey val="0"/>
                    <c:showVal val="1"/>
                    <c:showCatName val="0"/>
                    <c:showSerName val="0"/>
                    <c:showPercent val="0"/>
                    <c:showBubbleSize val="0"/>
                    <c:showLeaderLines val="1"/>
                </c:dLbls>
                <c:firstSliceAng val="0"/>
            </c:pieChart>
        </c:plotArea>
        <c:legend>
            <c:legendPos val="r"/>
            <c:overlay val="0"/>
        </c:legend>
        <c:plotVisOnly val="1"/>
        <c:dispBlanksAs val="gap"/>
        <c:showDLblsOverMax val="1"/>
    </c:chart>
    <c:txPr>
        <a:bodyPr/>
        <a:lstStyle/>
        <a:p>
            <a:pPr>
                <a:defRPr sz="1200">
                    <a:latin typeface="Calibri Light "/>
                </a:defRPr>
            </a:pPr>
            <a:endParaRPr lang="de-DE"/>
        </a:p>
    </c:txPr>
    <c:externalData r:id="rId1">
        <c:autoUpdate val="0"/>
    </c:externalData>
</c:chartSpace>
```

I have not compared these two XMLs in a detailed manner, there seem to be multiple changes, but I have found the `autoTitleDeleted` property on the new one as you've guessed. I will try the code piece you mentioned to see how it reflects, will keep here posted.

**Comment by karakanb** [comment]
Yeah, the piece of code you posted solved this issue, thank you very much. 

As a general question, do you have any idea what causes these specific differences between the original document and PowerPoint opened version? Even language has been modified to `de-DE`, even though I am using PowerPoint in English. What should be the best practice around these topics to make the resulting slides as much cross-platform as possible?

Thank you very much again for your quick reply and caring approach, amazing job with the package.

**Comment by scanny** [comment]
Well, the cross-compatibility is quite good, I'd say at least 99% maybe 99.9%. This particular item is a little bit of a backwater, only being operative in the case of single-series charts (which pie/donut charts always are). The ISO29500 spec is open to interpretation as far as defaults for things like this are concerned so perhaps that's why LibreOffice and PowerPoint did it different ways. My own opinion is that PowerPoint behavior should be the defacto spec where the spec is silent, but no one at LibreOffice asks me about that sort of thing :)

On the other hand, I think it would be a good improvement to `python-pptx` if we wrote the `c:autoTitleDeleted` element when we added a new chart. I'll leave this issue open as a request for that.
